File: Languages/go/low-level-design-lld/design-patterns/singleton.go
package main

import (
	"fmt"
	"sync"
)

// Singleton demonstrates thread-safe singleton using sync.Once.
// sync.Once guarantees the initialization function runs exactly once,
// even when called concurrently from multiple goroutines.

type database struct {
	connection string
}

var (
	dbInstance *database
	once       sync.Once
)

// GetDatabase returns the singleton database instance.
// The first call initializes the connection; subsequent calls return the same instance.
func GetDatabase() *database {
	once.Do(func() {
		fmt.Println("Creating database connection (this runs only once)")
		dbInstance = &database{connection: "postgres://localhost:5432/mydb"}
	})
	return dbInstance
}

func main() {
	// Demonstrate thread-safety: spawn multiple goroutines requesting the singleton
	var wg sync.WaitGroup

	for i := 0; i < 5; i++ {
		wg.Add(1)
		go func(id int) {
			defer wg.Done()
			db := GetDatabase()
			fmt.Printf("Goroutine %d got connection: %s\n", id, db.connection)
		}(i)
	}

	wg.Wait()

	// Verify same instance
	db1 := GetDatabase()
	db2 := GetDatabase()
	fmt.Printf("Same instance? %v\n", db1 == db2)
}

File: Languages/javascript/low-level-design-lld/design-patterns/singleton.mjs
/**
 * Singleton Design Pattern
 * Uses closure and module pattern for a single instance.
 */

class Database {
  constructor() {
    if (Database._instance) {
      return Database._instance;
    }
    this.connection = "PostgreSQL@localhost:5432";
    Database._instance = this;
  }

  query(sql) {
    return `Executing: ${sql}`;
  }

  static getInstance() {
    if (!Database._instance) {
      new Database();
    }
    return Database._instance;
  }
}

export { Database };

// Tests
import { test } from "node:test";
import assert from "node:assert/strict";

test("Singleton - same instance", () => {
  const db1 = Database.getInstance();
  const db2 = Database.getInstance();
  assert.strictEqual(db1, db2);
});

test("Singleton - correct connection", () => {
  const db = Database.getInstance();
  assert.strictEqual(db.connection, "PostgreSQL@localhost:5432");
});

test("Singleton - query works", () => {
  const db = Database.getInstance();
  assert.strictEqual(db.query("SELECT 1"), "Executing: SELECT 1");
});

test("Singleton - new returns same instance", () => {
  const db1 = new Database();
  const db2 = new Database();
  assert.strictEqual(db1, db2);
});
